resources/js/tableofcontent.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * Construct recurseData.toc as a JSON-like that contains each header's text,
+ * anchor, and their list of child headers.
+ * 
+ * References:
+ * https://stackoverflow.com/questions/187619/is-there-a-javascript-solution-to-generating-a-table-of-contents-for-a-page
+ * @param elem DOM element
+ * @param recurseData Object with fields: { toc: Object, count: int, parent: Object }
+ */
+function recurseChild(elem, recurseData) {
+    const { parent } = recurseData;
+
+    if (isHeaderElement(elem))
+    {
+        const headerData = createHeaderData(elem, recurseData);
+        recurseData.parent = insertHeaderData(parent, headerData);
+    }
+
+    Array.from(elem.children).forEach(child => recurseChild(child, recurseData));
+}
+
+function isHeaderElement(elem) {
+    return elem.nodeName.match(/^H\d+$/) && elem.textContent;
+}
+
+function createHeaderData(elem, recurseData) {
+    const headerText = elem.textContent.trim().replace(/"/g, '\\"');
+    const level = parseInt(elem.nodeName.substr(1));
+    const anchor = `kiwix-toc-${recurseData.count++}`;
+
+    const anchorElem = document.createElement("a");
+    anchorElem.id = anchor;
+
+    /* Wrap header content with something we can reference. */
+    elem.insertAdjacentElement("afterbegin", anchorElem);
+
+    return { headerEntry: { text: headerText, anchor, table: [] }, level };
+}
+
+function insertHeaderData(parent, headerData) {
+    const { headerEntry, level } = headerData;
+    const headerParentData = { parent: parent, table: headerEntry.table, level };
+
+    /* Recurse back until the correct level for insertion. */
+    while (parent.level !== 0 && level < parent.level)
+        parent = parent.parent;
+
+    /* Same level as parent meaning we should be sibling with parent. */
+    if (level === parent.level && parent.level !== 0)
+        headerParentData.parent = parent.parent;
+    headerParentData.parent.table.push(headerEntry);
+
+    return headerParentData;
+}
+
+function tocJSON()
+{
+    /* toc used to store constructed nested header object.
+       count used to uniquely identify each list item in toc.
+       parent used for toc construction
+    */
+    var recurseData = { toc: { url: window.location.href.replace(location.hash,""), table : []}, count: 0, parent: null };
+
+    /* parent field is the object's parent used for nest checking.
+       table field is the reference to the parent field's table array.
+       level field is the level of the parent field
+    */ 
+    recurseData.parent =  { parent: null, table: recurseData.toc.table, level: 0 };
+
+    if (document.body !== undefined) 
+        recurseChild(document.body, recurseData); 
+    return JSON.stringify(recurseData.toc);
+}
```

Comment:
I think the presence of two hierarchies (DOM and TOC) is a source of some confusion here. To the best of my understanding the DOM hierarchy is completely ignored while building the TOC hierarchy. Then the code could better be structured as follows:
1. Run depth first search (DFS) traversal of the DOM sending `<h*>` elements to a callback (or just build a list of those elements in DFS order).
2. Create the TOC hierarchy from the sequence of `<h*>` DOM elements (and at the same time create and insert the anchor elements).

Does it make (more) sense?

Reply:
Note that there are different variants of depth first search (pre-order, in-order, post-order) but the difference between them would only matter if we expect that a  `<h*>` element may contain other nested `<h*>` elements in its subtree. I think that pre-order DFS should be used.

Reply:
Done